azure-kusto-data/azure/kusto/data/_token_providers.py

```diff
@@ -218,9 +218,10 @@ def _get_token_from_cache_impl(self) -> dict:
 class CallbackTokenProvider(TokenProviderBase):
     """Callback Token Provider generates a token based on a callback function provided by the caller"""
 
-    def __init__(self, token_callback: Callable[[], str]):
+    def __init__(self, token_callback: Optional[Callable[[], str]], async_token_callback: Optional[Callable[[], Coroutine[None, None, str]]] = None):

```

Comment:
This is an odd constructor.
I must provide a sync callable if I want to use the async one.
I can't think of a more reasonable solution. Perhaps __init__(self, tokenCallback, isAsync = false)

Reply:
You can pass None for the sync callback

Reply:
Either way, that's probably not the interface we should be worried about since it's not user-facing

`KustoConnectionStringBuilder.with_token_provider` is currently the same, but maybe we can split it to `with_token_provider` and `with_async_token_provider`?
It would prevent the user from supplying both, but I don't think that's a real usecase

Reply:
makes more sense.
If you do init tokenProvider with Nones make both mandatory and use named arguments in the using code

Reply:
Is this what you meant @yogilad ?